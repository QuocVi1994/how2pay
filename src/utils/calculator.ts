import * as D from "dinero.js";

const RMBCurrency = { code: "RMB", base: 10, exponent: 2 };

export type RMB = D.Dinero<number>;

export function rmb(amount: number = 0): RMB {
  return D.dinero({
    amount: Math.round(amount * 100),
    currency: RMBCurrency,
  });
}

export function couponCalculator(goods: IGoods[], couponMap: ICouponMap) {
  const findCouponsOriginal = () => {
    const names = goods.flatMap((e) => e.coupons || []);
    return [...new Set(names)]
      .map((name) => couponMap[name])
      .filter((c) => (c.order ?? 0) <= 1) // 基于原价的优惠
      .filter((c) => c.times && c.times > 0) // 次数得够
      .sort((a, b) => (a.order || 3) - (b.order || 3));
  };

  const findCouponsNotOriginal = () => {
    const names = goods.flatMap((e) => e.coupons || []);
    return [...new Set(names)]
      .map((name) => couponMap[name])
      .filter((c) => (c.order ?? 0) > 1) // 基于到手价的优惠
      .filter((c) => c.times && c.times > 0) // 次数得够
      .sort((a, b) => (a.order || 3) - (b.order || 3));
  };

  // 记录每轮的匹配情况
  const turnMatch: { coupon: string; save: RMB }[] = [];

  // 保存商品的总价优惠
  const saveMap = {} as Record<string, RMB>;

  // 第一轮，按 总价 计算优惠
  findCouponsOriginal().forEach((coupon) => {
    const { name } = coupon;
    // 找出能用这个优惠券的商品
    const goodsWithCou = goods.filter((g) => g.coupons?.includes(name));
    // 计算能用该优惠券的总价
    const sum = sumGoods(goodsWithCou);
    // 计算用了这个优惠券能省多少钱
    const save = calCoupon(coupon, sum);
    // 计算每个商品的优惠价格
    const allocate = D.allocate(
      save,
      goodsWithCou.map((g) => g.price)
    );

    goodsWithCou.forEach((good, index) => {
      const key = good.name;
      if (!saveMap[key]) saveMap[key] = allocate[index];
      else saveMap[key] = D.add(saveMap[key], allocate[index]);
    });

    // 记录使用的每个优惠券 和 每个优惠券能省的钱
    if (D.greaterThan(save, rmb(0))) {
      turnMatch.push({ coupon: name, save });
    }
  });

  // 第二轮，按 到手价 计算优惠
  findCouponsNotOriginal().forEach((coupon) => {
    const { name } = coupon;
    // 找出能用这个优惠券的商品
    const goodsWithCou = goods.filter((g) => g.coupons?.includes(name));
    // 计算能用该优惠券的总价
    const sum = sumGoods(goodsWithCou);
    // 计算已经优惠了多少
    const saleSum = D.add(
      sumList(goodsWithCou.map((g) => g.shopSale || 0)),
      goodsWithCou
        .map((g) => saveMap[g.name] || rmb(0))
        .reduce((p, n) => D.add(p, n), rmb(0))
    );
    // 计算用了这个优惠券能省多少钱
    const save = calCoupon(coupon, D.subtract(sum, saleSum));

    // 记录使用的每个优惠券 和 每个优惠券能省的钱
    if (D.greaterThan(save, rmb(0))) {
      turnMatch.push({ coupon: name, save });
    }
  });

  // 原始总价
  const original = sumGoods(goods);

  // 店铺优惠
  const shopSale = sumList(goods.map((g) => g.shopSale || 0));

  // 所有优惠券的优惠价
  const couponTotal = sumRMB(turnMatch.map((e) => e.save));

  // 各种优惠的总和
  const discount = sumRMB([shopSale, couponTotal]);

  // 已付定金
  const deposit = sumList(goods.map((g) => g.deposit || 0));

  const payables = D.subtract(D.subtract(original, deposit), discount);

  return {
    original,
    discount,
    shopSale,
    couponTotal,
    deposit,
    payables,
    turnMatch,
  };
}

// 计算这个价格用这个优惠券，能优惠多少
function calCoupon(coupon: ICoupon, price: RMB): RMB {
  let save = rmb(0);

  const { type } = coupon;
  const target = rmb(coupon.target);
  const noTarget = !coupon.target || coupon.target <= 0;

  // 满减
  if (
    type === "target" &&
    !noTarget &&
    D.greaterThanOrEqual(price, target) &&
    coupon.times &&
    coupon.times > 0
  ) {
    // 一个优惠券允许多次使用
    if (coupon.repeat) {
      const temp = D.toUnit(price) / D.toUnit(target);
      const times = Math.min(coupon.times, Math.floor(temp));
      save = D.multiply(rmb(coupon.reduce), times);
      coupon.times -= times; // 优惠券减去用的次数
    }
    // 优惠券只能用一次
    else {
      save = rmb(coupon.reduce || 0);
      coupon.times -= 1;
    }
  }
  // 打折
  else if (type === "sale" && coupon.sale && coupon.sale > 0) {
    // 万一有些优惠券还需要到达多少钱才能打折
    if (noTarget || D.greaterThanOrEqual(price, target)) {
      // 这里淘宝的计算是去掉小数点后两位
      // 所以要转成number计算
      save = rmb(D.toUnit(price) * (1 - coupon.sale));
    }
  }
  // 并列的
  else if (type === "group" && coupon.group?.length) {
    // 分组一定得是满减的，按减的价格最高来降序
    const groups = coupon.group
      .filter((e) => e.type === "target" && e.target && e.reduce)
      .sort((a, b) => a.reduce! - b.reduce!);

    for (const cou of groups) {
      save = calCoupon(cou, price);
      if (D.greaterThan(save, rmb(0))) {
        break;
      }
    }
  }

  return save;
}

// 原价
function sumGoods(goodsList: IGoods[]): RMB {
  return goodsList.reduce((sum, goods) => D.add(sum, rmb(goods.price)), rmb(0));
}

function sumList(list: number[]): RMB {
  return list.reduce((p, n) => D.add(p, rmb(n)), rmb(0));
}

function sumRMB(list: RMB[]): RMB {
  return list.reduce((p, n) => D.add(p, n), rmb(0));
}
