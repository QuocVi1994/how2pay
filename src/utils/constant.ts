export const tbCoupons: ICoupon[] = [
  {
    name: "满300减50",
    type: "target",
    order: 0,
    times: 99999,
    repeat: true,
    target: 300,
    reduce: 50,
  },
  {
    name: "购物券1200-110",
    type: "target",
    order: 1,
    times: 1,
    repeat: false,
    target: 1200,
    reduce: 110,
  },
  {
    name: "88vip打折",
    type: "sale",
    order: 3,
    times: 9999,
    repeat: false,
    target: 0,
    sale: 0.95,
  },
  {
    name: "88vip优惠套餐",
    type: "group",
    order: 4,
    times: 99,
    repeat: false,
    group: [
      {
        name: "6800-500",
        type: "target",
        times: 1,
        order: 0,
        repeat: false,
        target: 6800,
        reduce: 500,
      },
      {
        name: "3800-300",
        type: "target",
        times: 1,
        order: 0,
        repeat: false,
        target: 3800,
        reduce: 300,
      },
      {
        name: "1000-50",
        type: "target",
        times: 1,
        order: 0,
        repeat: false,
        target: 1000,
        reduce: 50,
      },
    ],
  },
];
