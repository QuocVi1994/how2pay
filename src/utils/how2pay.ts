import * as D from "dinero.js";
import { cloneDeep } from "lodash-es";
import { couponCalculator, rmb } from "./calculator";

export function how2pay(goods: IGoods[], coupons: ICoupon[]) {
  const couponMap = Object.fromEntries(coupons.map((c) => [c.name, c]));

  let minPaid = rmb(1e9); // 10亿
  let groupsSave: IGoods[][] = [];
  let resultSave: ReturnType<typeof couponCalculator>[] = [];

  for (const groups of genGoodsGroup(goods)) {
    const map = cloneDeep(couponMap);

    let payables = rmb(0);

    const results = groups.map((group) => {
      const result = couponCalculator(group, map);
      payables = D.add(payables, result.payables);
      return result;
    });

    if (D.greaterThan(minPaid, payables)) {
      minPaid = payables;
      resultSave = results;
      groupsSave = groups;
    }

    console.log(D.toUnit(payables));
  }

  return {
    minPaid,
    resultSave,
    groupsSave,
  };
}

export function* genFullSort<T>(list: T[]): Generator<T[]> {
  const len = list.length;
  if (len === 1) {
    yield list;
  } else if (len === 2) {
    yield [list[0], list[1]];
    yield [list[1], list[0]];
  } else {
    for (let i = 0; i < len; i++) {
      const temp = [...list];
      const pre = temp[i];
      temp.splice(i, 1);

      for (const ret of genFullSort(temp)) {
        yield [pre, ...ret];
      }
    }
  }
}

export function* genGoodsGroup(goods: IGoods[]): Generator<IGoods[][]> {
  // 拿到所有的优惠券
  const allCoupons = [...new Set(goods.flatMap((g) => g.coupons || []))];

  const findIndex = (good: IGoods, coupons: string[]) => {
    return (
      good.coupons?.reduce((p, g) => {
        return Math.max(p, coupons.indexOf(g));
      }, -1) || -1
    );
  };

  // 不分组
  yield [[...goods]];

  // 按优惠券的来优先分组
  for (const coupons of genFullSort(allCoupons)) {
    const groups: IGoods[][] = [...coupons.map(() => []), []];

    goods.forEach((good) => {
      let index = findIndex(good, coupons);
      index = index === -1 ? coupons.length : index;
      groups[index].push(good);
    });

    yield groups.filter((g) => g.length > 0);
  }
}
