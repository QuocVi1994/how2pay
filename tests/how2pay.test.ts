import * as D from "dinero.js";
import { expect, it } from "vitest";
import { cloneDeep } from "lodash-es";
import { couponCalculator } from "../src/utils/calculator";
import { genFullSort, how2pay } from "../src/utils/how2pay";

const coupons: ICoupon[] = [
  {
    name: "满300减50",
    type: "target",
    order: 0,
    times: 99999,
    repeat: true,
    target: 300,
    reduce: 50,
  },
  {
    name: "购物券1200-110",
    type: "target",
    order: 1,
    times: 1,
    repeat: false,
    target: 1200,
    reduce: 110,
  },
  {
    name: "88vip打折",
    type: "sale",
    order: 3,
    times: 9999,
    repeat: false,
    target: 0,
    sale: 0.95,
  },
  {
    name: "88vip优惠套餐",
    type: "group",
    order: 4,
    times: 99,
    repeat: false,
    group: [
      {
        name: "6800-500",
        type: "target",
        times: 1,
        order: 0,
        repeat: false,
        target: 6800,
        reduce: 500,
      },
      {
        name: "3800-300",
        type: "target",
        times: 1,
        order: 0,
        repeat: false,
        target: 3800,
        reduce: 300,
      },
      {
        name: "1000-50",
        type: "target",
        times: 1,
        order: 0,
        repeat: false,
        target: 1000,
        reduce: 50,
      },
    ],
  },
];

const goods: IGoods[] = [
  {
    name: "bose",
    price: 1999,
    deposit: 100,
    shopSale: 0,
    coupons: ["满300减50", "88vip打折", "88vip优惠套餐"],
  },
  {
    name: "伊夫泉",
    price: 900,
    deposit: 50,
    shopSale: 451,
    coupons: ["满300减50", "购物券1200-110", "88vip优惠套餐"],
  },
  {
    name: "mac粉底液",
    price: 340,
    deposit: 50,
    shopSale: 0,
    coupons: ["88vip优惠套餐"],
  },
  {
    name: "修丽可",
    price: 595,
    deposit: 50,
    shopSale: 10,
    coupons: ["购物券1200-110", "88vip优惠套餐"],
  },
  {
    name: "自然堂面膜",
    price: 398,
    deposit: 40,
    shopSale: 150,
    coupons: ["满300减50", "88vip优惠套餐"],
  },
];

it("couponCalculator", () => {
  const map = cloneDeep(Object.fromEntries(coupons.map((e) => [e.name, e])));
  const result = couponCalculator(goods, map);

  expect(helpDebug(result)).toMatchSnapshot();
});

it("genGroups", () => {
  let i = 0;
  for (const group of genFullSort([1, 2, 3])) {
    console.log(++i, "\t", group);
  }
});

it("how2pay", () => {
  const result = how2pay(goods, cloneDeep(coupons));

  expect(helpDebug(result)).toMatchSnapshot();
});

function helpDebug(input: any): any {
  if (Array.isArray(input)) {
    return input.map((e) => helpDebug(e));
  } else if (typeof input === "object") {
    if (input.calculator && input.create && input.toJSON) {
      return D.toUnit(input);
    } else {
      return Object.fromEntries(
        Object.entries(input).map(([key, value]: any) => [
          key,
          helpDebug(value),
        ])
      );
    }
  }

  return input;
}
