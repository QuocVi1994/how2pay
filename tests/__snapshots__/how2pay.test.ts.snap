// Vitest Snapshot v1

exports[`couponCalculator 1`] = `
{
  "couponTotal": 744.79,
  "deposit": 290,
  "discount": 1355.79,
  "original": 4232,
  "payables": 2586.21,
  "shopSale": 611,
  "turnMatch": [
    {
      "coupon": "满300减50",
      "save": 500,
    },
    {
      "coupon": "购物券1200-110",
      "save": 110,
    },
    {
      "coupon": "88vip打折",
      "save": 84.79,
    },
    {
      "coupon": "88vip优惠套餐",
      "save": 50,
    },
  ],
}
`;

exports[`how2pay 1`] = `
{
  "groupsSave": [
    [
      {
        "coupons": [
          "满300减50",
          "88vip打折",
          "88vip优惠套餐",
        ],
        "deposit": 100,
        "name": "bose",
        "price": 1999,
        "shopSale": 0,
      },
      {
        "coupons": [
          "88vip优惠套餐",
        ],
        "deposit": 50,
        "name": "mac粉底液",
        "price": 340,
        "shopSale": 0,
      },
      {
        "coupons": [
          "满300减50",
          "88vip优惠套餐",
        ],
        "deposit": 40,
        "name": "自然堂面膜",
        "price": 398,
        "shopSale": 150,
      },
    ],
    [
      {
        "coupons": [
          "满300减50",
          "购物券1200-110",
          "88vip优惠套餐",
        ],
        "deposit": 50,
        "name": "伊夫泉",
        "price": 900,
        "shopSale": 451,
      },
      {
        "coupons": [
          "购物券1200-110",
          "88vip优惠套餐",
        ],
        "deposit": 50,
        "name": "修丽可",
        "price": 595,
        "shopSale": 10,
      },
    ],
  ],
  "minPaid": 2585.64,
  "resultSave": [
    {
      "couponTotal": 485.36,
      "deposit": 190,
      "discount": 635.36,
      "original": 2737,
      "payables": 1911.64,
      "shopSale": 150,
      "turnMatch": [
        {
          "coupon": "满300减50",
          "save": 350,
        },
        {
          "coupon": "88vip打折",
          "save": 85.36,
        },
        {
          "coupon": "88vip优惠套餐",
          "save": 50,
        },
      ],
    },
    {
      "couponTotal": 260,
      "deposit": 100,
      "discount": 721,
      "original": 1495,
      "payables": 674,
      "shopSale": 461,
      "turnMatch": [
        {
          "coupon": "满300减50",
          "save": 150,
        },
        {
          "coupon": "购物券1200-110",
          "save": 110,
        },
      ],
    },
  ],
}
`;
